cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py

```diff
@@ -215,32 +230,35 @@ def setup_leader_authentication():
     # eg:
     # {'/root/cdk/serviceaccount.key': 'RSA:2471731...'}
     charms.leadership.leader_set(leader_data)
-
     set_state('authentication.setup')
 
 
 @when_not('leadership.is_leader')
-@when_not('authentication.setup')
 def setup_non_leader_authentication():
-    api_opts = FlagManager('kube-apiserver')
-    controller_opts = FlagManager('kube-controller-manager')
 
     service_key = '/root/cdk/serviceaccount.key'
     basic_auth = '/root/cdk/basic_auth.csv'
     known_tokens = '/root/cdk/known_tokens.csv'
 
-    hookenv.status_set('maintenance', 'Rendering authentication templates.')
-
     keys = [service_key, basic_auth, known_tokens]
     if not get_keys_from_leader(keys):
         # the keys were not retrieved. Non-leaders have to retry.
         return
 
+    if not any_file_changed(keys) and is_state('authentication.setup'):
+        # No change detected and we have already setup the authentication
+        return
+
+    hookenv.status_set('maintenance', 'Rendering authentication templates.')
+    api_opts = FlagManager('kube-apiserver')
     api_opts.add('basic-auth-file', basic_auth)
     api_opts.add('token-auth-file', known_tokens)
     api_opts.add('service-account-key-file', service_key)
+
+    controller_opts = FlagManager('kube-controller-manager')
     controller_opts.add('service-account-private-key-file', service_key)
 
+    remove_state('kubernetes-master.components.started')
```

Comment:
Yes we can do that.

On the leader we have two triggers that cause a reconfiguration of the authentication 1) charm upgrade 2) password change. In both triggers we do 
```
    remove_state('authentication.setup')
    remove_state('kubernetes-master.components.started')
```

On the non-leaders the reconfig of the authorisation is triggered after detecting a change on the leader. This is why I had the remove state there.

To make it homogeneous I added the remove state on the `setup_leader_authentication` and removed it from the upgrade and change password triggers.